src/eexprs.adoc

```diff
@@ -0,0 +1,100 @@
+This reflects the idea that macros behave like new syntactic forms, with this
+entire character sequence determining the syntax that follows inside the
+expression.
+
+Following the opening macro reference and whitespace, E-expressions follow
+S-expression tokenization and whitespace rules.
+The remaining elements are arguments supplying inputs to the macro,
+each one either _individual_ or _grouped_.
+The syntax that can appear in each argument position is constrained by the
+macro’s signature.
+
+The number of argument elements (that is, the invocation’s actual arity)
+must be equal to or greater than the macro’s minimum arity, and at most its
+maximum arity, when one exists.
+In other words, an E-expression must contain one element for each required
+parameter, followed by optional elements for the remaining optional parameters.
+
+// TODO links for arity, optional parameters
+// TODO base type? base shape?  base form?  encoding?
+
+The syntax of each argument is defined by the associated parameter’s cardinality
+and type.
+The cardinality determines whether an argument group can be used to collect a
+series of individual arguments, and the type determines the syntax of an
+individual argument, grouped or not.
+
+The parameter type constrains the syntax of an individual argument as follows:
+
+* For tagged types, an individual argument must be a single E-expression or any
+  datum (which may contain nested E-expressions).
+* For primitive types, an individual argument must be a non-null, non-annotated
+  datum of the corresponding concrete type, within the range accepted by the
+  primitive type.
+* For macro types, an individual argument must be an unannotated S-expression
+  containing arguments acceptable to that macro’s signature.
+  Here, the S-expression is implicitly converted to the equivalent E-expression.
+
+The parameter cardinality constrains the syntax of the overall argument
+element(s), particularly whether a group is allowed.
+In text E-expressions, argument groups are delimited using the special syntax
+`(: …)` where the colon is followed by whitespace or `)` instead of a macro
+reference.
+Each element of an argument group must fit the same syntax rules as for an
+individual argument.
+
+* A `!` parameter accepts only a single individual argument.
+* A `?` parameter accepts either a single individual argument
+  or an empty argument group `(:)`.
+* A `*` or `+` parameter accepts either a single individual argument,
+  or an argument group `(: …)`.
+* A rest parameter captures all remaining arguments of the E-expression,
+  each of which must match the individual argument syntax.
+  For a `...+` parameter there must be at least one such argument.
+
+// TODO clarify whether a `+` group must contain at least one element.
+
+The rules above determine whether an E-expression is _well-formed_.
+Any violation of the above constraints must signal a syntax error when the
+E-expression is parsed.
```

Comment:
Does this imply anything about the eagerness of the parsing? Or whether an Ion reader may skip parsing some data in a stream (e.g. by skipping past a container)?

Reply:
I tried to avoid any such implication here, since I don't think we know exactly how that should work.   I think that gap is part of #307 noted in the next paragraph.